### ISSUE TITLE
Revamp sharktank packaging and releases

### ISSUE THREAD

**Issue by ScottTodd** [nod-ai/shark-ai#294]
See also https://github.com/nod-ai/SHARK-Platform/issues/130 for shortfin.

Development/nightly releases of the shortfin package are being published to https://github.com/nod-ai/SHARK-Platform/releases/tag/dev-wheels using https://github.com/nod-ai/SHARK-Platform/blob/main/.github/workflows/build_packages.yml and https://github.com/nod-ai/SHARK-Platform/blob/main/shortfin/build_tools/build_linux_package.sh.

The sharktank package should be published to the same location(s) for the same set of Python versions and operating systems. Each package should follow the same versioning scheme (though we may release minor/patch versions of each package independently). cc @marbre for versioning considerations

We still need to decide concretely how these packages will be published to pypi. The existing https://pypi.org/project/sharktank/ and https://pypi.org/project/shortfin/ pages have a few dev releases that we can yank in favor of more recent packages with whatever version scheme we decide on. If we settle on a good stable / pre-release / nightly workflow with pypi, we can maybe avoid pushing to github releases.

The metadata and configuration in https://github.com/nod-ai/SHARK-Platform/blob/main/sharktank/setup.py should be reexamined, as should the requirements (dev/test/install/etc.). It sounded like @stellaraccident wanted to take a dep on shortfin from sharktank too.

**Comment by ScottTodd** [comment]
With https://github.com/nod-ai/SHARK-Platform/pull/404 sharktank wheels are now published nightly to https://github.com/nod-ai/SHARK-Platform/releases/tag/dev-wheels

They can be installed with:
```
pip install sharktank-f https://github.com/nod-ai/SHARK-Platform/releases/expanded_assets/dev-wheels
```

More work is needed to get the wheels usable

> Note that the 250KB Python wheel can be _installed_ standalone, but actually _using_ it will need version bumps across projects and more work. In particular, I hit
> 
> 1. Using iree-turbine from pypi:
>    ```
>    from sharktank import ops
>    ...
>    
>    ImportError: cannot import name 'dynamic_dim' from 'torch.export' (/home/nod/dev/scratch/shark-platform/3.11.venv/lib/python3.11/site-packages/torch/export/__init__.py)
>    ```
>      
>    That may be fixed by [Remove deprecated constraints and dynamic dim support with test fixes iree-org/iree-turbine#233](https://github.com/iree-org/iree-turbine/pull/233)
> 2. After installing iree-turbine from source:
>    ```
>    from sharktank import ops
>    ...
>    
>        from gguf import GGUFReader, GGUFValueType
>    ModuleNotFoundError: No module named 'gguf'
>    ```
> 
> Full logs here: https://gist.github.com/ScottTodd/4cd0c343a92375e7fbdaaf37c38bfc06

Some ideas:

- [ ] Add more packages as dependencies if they are base functionality (e.g. `gguf`)
- [ ] Make more packages optional (error when trying to use specific features, not at import time)
- [ ] Push a new release of iree-turbine to pypi
- [ ] Put together dev instructions for getting nightly versions of all shark+iree packages in the dependency graph

Also need to test a variety of python versions, torch versions, etc.

**Comment by ScottTodd** [comment]
sharktank packaging changes are done now. More restructuring is planned around some dependencies like `torch` and we'll also want to add the `sharktank` package back to the `shark-ai` meta package once it stabilizes more.
